### PR TITLE
8269886: Inaccurate error message for compressed hprof test

### PR DESCRIPTION
**Pull request by linzang**
The current implementation of hprof Reader for testing always prompts "Can not decompress the compressed hprof file" when there is error for testing gzipped heap dump. This is inaccurate if the gzipped file was decompressed successfully but the hprof file format is incorrect. So the inaccurate error message could be misleading for issue analysis.

This trivial PR refine the error message by simply print "Can not get stack trace from the compressed hprof file", the underlying exception from GZIPInputStream() or HprofReader() would give accurate error info.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269886](https://bugs.openjdk.java.net/browse/JDK-8269886): Inaccurate error message for compressed hprof test

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4685/head:pull/4685` \
`$ git checkout pull/4685`

Update a local copy of the PR: \
`$ git checkout pull/4685` \
`$ git pull https://git.openjdk.java.net/jdk pull/4685/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4685`

View PR using the GUI difftool: \
`$ git pr show -t 4685`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4685.diff">https://git.openjdk.java.net/jdk/pull/4685.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
The existing message is a generic message covering the general operation of the whole try block. It seems far more appropriate than your new message, which only seems to apply to the final step. ???

David

**Comment by linzang** [comment]
Hi David,
    Thanks for review and comments.

> The existing message is a generic message covering the general operation of the whole try block. It seems far more appropriate than your new message, which only seems to apply to the final step. ???

IMO, the whole method named getStack() is used to get the stack trace for the hprof file.  And the try block could throw exceptions by serveral reasons as decribed, so IMHO printing "Can not decompress the compressed hprof file" is not accurate. The proposed change is to print a generic error message that the getStack() fail when processing compressed hprof.  And the extra 'e' argument in the changed line would give detailed reason.  

So may be change error message to "Can not process the compressed hprof file" is better? what do you think? 

BRs,
Lin

**Comment by linzang** [comment]
Hi @dholmes-ora， 
     Thanks for your suggestion! I have update the PR which split the try block into two seperate ones to process different exceptions.

BRs,
Lin

**Review by dholmes-ora — APPROVED** [review]
These changes are fine in themselves.

This code is not very well written though - the file streams do not get closed and the decompressed output file doesn't seem to get deleted when I would expect.

Thanks,
David

**Comment by linzang** [comment]
Hi David，
Thanks for reviewing!
 
> This code is not very well written though - the file streams do not get closed and the decompressed output file doesn't seem to get deleted when I would expect.

Agree. I will create a seperate issue to fix it.

BRs,
Lin

**Comment by plummercj** [comment]
You need 2 reviews. Please don't integrate yet. I'll provide more comments shortly.

**Comment by plummercj** [comment]
After decompressing, no error or exception is thrown if MAGIC_NUMBER is not found. It just silently fails to print the stack trace.

Please use "cannot" instead of "can not". Both are grammatically correct, but "cannot" is what is normally used.

**Comment by linzang** [comment]
Hi @plummercj , 

Thanks for reviewing,  I have updated the PR based on your comments, and remove unreachable code introduced by the updated code.

BRs,
Lin

File: test/lib/jdk/test/lib/hprof/parser/Reader.java [hunk 1]
```diff
@@ -168,16 +172,18 @@ public static String getStack(String heapFile, int debugLevel)
                             = new HprofReader(deCompressedFile, in2, dumpNumber,
                                               true, debugLevel);
                         r.read();
+                        out.delete();
```
Comment (dholmes-ora) [inline test/lib/jdk/test/lib/hprof/parser/Reader.java:175]: I thought this was going to handled by the separate RFE to fix the general logic? And you don't delete before throwing below. There should be a finally clause to handle deletion consistently.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Either deleting the out file needs to be handled correctly in all cases or it should left as-is and fixed in a different RFE that also handles stream closing *try-with-resources?) etc. to clean up the code.

Otherwise the actual try/catch changes seem okay.

Thanks,
David

**Comment by linzang** [comment]
Hi @dholmes-ora 

> Either deleting the out file needs to be handled correctly in all cases or it should left as-is and fixed in a different RFE that also handles stream closing *try-with-resources?) etc. to clean up the code.

I have created an issue in JBS (https://bugs.openjdk.java.net/browse/JDK-8269909) to use the try-with-resource for stream closing, and will submit a PR shortly. 

For out.delete(), I prefer to fix it in this PR as the introduction of throwing exception in my last update makes the original out.delete() unreachable, so I have to add logic to handle it.  I will update this PR with a finally block in a minute. 

Thanks,
Lin

**Review by dholmes-ora — APPROVED** [review]
I would have added a single try/finally block that encompasses all the code after:

File out = new File(deCompressedFile);

but what you have also does the job.

Thanks,
David

**Comment by plummercj** [comment]
There is still an inconsistency exception handling. For the outer bad MAGIC_NUMBER handling you have:

186                 throw new IOException("Unrecognized magic number: " + i);

But the decompressed handling results in:

178                         throw new IOException("Unrecognized magic number of decompressed data: " + i);

Which is caught and wrapped by:

181                     throw new IOException("Cannot get stack trace from the compressed hprof file", e);

Also, nothing is wrapping exceptions thrown by the following code:

143                 HprofReader r
144                     = new HprofReader(heapFile, in, dumpNumber,
145                                       true, debugLevel);
146                 r.read();
147                 return r.printStackTraces();

But any exception thrown by the following code is wrapped and rethrown at line 181:

172                         HprofReader r
173                             = new HprofReader(deCompressedFile, in2, dumpNumber,
174                                               true, debugLevel);
175                         r.read();
176                         return r.printStackTraces();

**Comment by linzang** [comment]
Hi @plummercj,
    Thanks for point it out! I think the throw at line 181 could be removed and the `new = HprofReader(); r.read()` code should be in try block consistently in both situation (compressed/uncompressed). I will make the change.
-Lin

**Comment by linzang** [comment]
Hi @plummercj,
    I just found that maybe there is no need to catch exception from `HprofReader` in this method, just keep it as what it originally did -  throw any exception from HprofReader directly. The information from HprofReader's exception is enough for root cause analysis in this case.
    I will test it and make the change.

BRs,
Lin

File: test/lib/jdk/test/lib/hprof/parser/Reader.java [hunk 2]
```diff
@@ -169,15 +174,15 @@ public static String getStack(String heapFile, int debugLevel)
                                               true, debugLevel);
                         r.read();
                         return r.printStackTraces();
+                    } else {
+                        throw new IOException("Unrecognized magic number of decompressed data: " + i);
```
Comment (plummercj) [inline test/lib/jdk/test/lib/hprof/parser/Reader.java:178]: "of decompressed data" just doesn't sound right. I'd suggest either removing it, or maybe instead say "...found in decompressed data".

**Review by plummercj — CHANGES_REQUESTED** [review]
Looks good other than cleaning up the exception message.

**Comment by linzang** [comment]
I will mark this PR with integrate. 
Thanks Chris and David for your reviewing!

BRs,
Lin

**Review by dholmes-ora — APPROVED** [review]
LGTM.

**Review by plummercj — APPROVED** [review]

**Pull request closed without merging**